File: 08_greedy_algorithms/python/greedy.py
states_needed = {"mt", "wa", "or", "id", "nv", "ut", "ca", "az"}

stations = {}
stations["kone"] = {"id", "nv", "ut"}
stations["ktwo"] = {"wa", "id", "mt"}
stations["kthree"] = {"or", "nv", "ca"}
stations["kfour"] = {"nv", "ut"}
stations["kfive"] = {"ca", "az"}

final_stations = set()

if __name__ == '__main__':
    while states_needed:
        best_station = None
        states_covered = set()
        for station, states in stations.items():
            covered = states_needed.intersection(states)  # states_needed & states
            if len(covered) > len(states_covered):
                best_station = station
                states_covered = covered

        states_needed -= states_covered
        final_stations.add(best_station)

    print(final_stations)

File: 08_greedy_algorithms/go/greedy.go
package main

import "fmt"

func main() {
	statesNeeded := []string{"mt", "wa", "or", "id", "nv", "ut", "ca", "az"}

	var stations = make(map[string][]string)
	stations["kone"] = []string{"id", "nv", "ut"}
	stations["ktwo"] = []string{"wa", "id", "mt"}
	stations["kthree"] = []string{"or", "nv", "ca"}
	stations["kfour"] = []string{"nv", "ut"}
	stations["kfive"] = []string{"ca", "az"}

	stationsKey := []string{"kone", "ktwo", "kthree", "kfour", "kfive"}

	var finalStations []string

	for len(statesNeeded) > 0 {
		var bestStation string
		var statesCovered []string

		for _, station := range stationsKey {
			states := stations[station]
			var covered = intersection(statesNeeded, states)
			if len(covered) > len(statesCovered) {
				bestStation = station
				statesCovered = covered
			}

		}
		statesNeeded = difference(statesNeeded, statesCovered)
		finalStations = append(finalStations, bestStation)
	}
	fmt.Println(finalStations)
}

func intersection(a, b []string) []string {
	var result []string

	for _, val := range a {
		if contains(b, val) {
			result = append(result, val)
		}
	}

	return result
}

func contains(a []string, x string) bool {
	for _, n := range a {
		if x == n {
			return true
		}
	}

	return false
}

func difference(a, b []string) []string {
	var result []string

	for _, val := range a {
		if !contains(b, val) {
			result = append(result, val)
		}
	}

	return result
}